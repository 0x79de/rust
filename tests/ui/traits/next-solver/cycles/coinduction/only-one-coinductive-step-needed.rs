//@ revisions: current next
//@ ignore-compare-mode-next-solver (explicit revisions)
//@[next] compile-flags: -Znext-solver
//@[next] check-pass

// #136824 changed cycles to be coinductive if they have at least
// one productive step, causing this test to pass with the new solver.

struct Foo<T>(<Foo<T> as Trait>::Assoc);
//[current]~^ ERROR overflow evaluating the requirement `Foo<T>: SendIndir`

trait SendIndir {}
impl<T: Send> SendIndir for T {}

trait Trait {
    type Assoc;
}
impl<T: SendIndir> Trait for T {
    type Assoc = ();
}

fn is_send<T: Send>() {}

fn main() {
    is_send::<Foo<u32>>();
}
