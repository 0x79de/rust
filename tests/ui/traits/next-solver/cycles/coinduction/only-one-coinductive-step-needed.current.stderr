error[E0275]: overflow evaluating the requirement `Foo<T>: SendIndir`
  --> $DIR/only-one-coinductive-step-needed.rs:9:15
   |
LL | struct Foo<T>(<Foo<T> as Trait>::Assoc);
   |               ^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: required for `Foo<T>` to implement `Trait`
  --> $DIR/only-one-coinductive-step-needed.rs:18:20
   |
LL | impl<T: SendIndir> Trait for T {
   |         ---------  ^^^^^     ^
   |         |
   |         unsatisfied trait bound introduced here

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0275`.
